### ISSUE TITLE
send ktls offload can't work on latest openssl

### ISSUE THREAD

**Issue by gaojunhao** [openssl/openssl#14951]
hi all
I use the latest nginx compiling with openssl, then find one problem that I use openssl(commit id: ac093b3fe6ba3f21e874a25ddecc7c1b4dff2765), send ktls offload can work, but use the latest openssl(commit id: 1c0eede9827b0962f1d752fa4ab5d436fa039da4), send ktls offload can't work.

My setting as follows:
centos kernel: 5.10
network card: mellnox connect-6dx
nginx version: latest code of https://github.com/nginx/nginx.git
openssl version: code from https://github.com/openssl/openssl.git, work commit id: ac093b3fe6ba3f21e874a25ddecc7c1b4dff2765, not work commit id: 1c0eede9827b0962f1d752fa4ab5d436fa039da4
gcc version: 7.5

openssl compile:
cd openssl; ./config; make -j32

nginx compile:
cd nginx; ./auto/configure --with-openssl=/path/to/openssl --with-debug --with-http_ssl_module --with-openssl-opt="enable-ktls -DOPENSSL_LINUX_TLS -g3"

nginx.conf:
server {
        listen 443 ssl default_server;
        server_name localhost;
        client_max_body_size 5G; 
        include /etc/nginx/default.d/*.conf;
        ssl_certificate /usr/local/nginx/conf/cert.pem;
        ssl_certificate_key /usr/local/nginx/conf/key.pem;
        ssl_ciphers ECDHE-RSA-AES128-GCM-SHA256;
        ssl_protocols TLSv1.2;
        location / { 
                root html;
                index index.html;
        }   
        #error_page 405 /uploadtest;
        error_page 404 /404.html;
        location = /40x.html {
                root html;
        }   
        error_page 500 502 503 504 /50x.html;
        location = /50x.html {
                root html;
        }   
        location = /test.img {
                root html;
        }   
}

then add one test.img into /usr/local/nginx/html
dd if=/dev/zero of=/usr/local/nginx/html/test.img bs=1M count=2048

then enable network card tx offload
ethtool -K net3 tls-hw-tx-offload on

then use curl command running in another machine:
curl -k https://nginx_server_ip:443/test.img -O

then tcpdump network interface. find the packet context is cipher context on nginx server, if I use send ktls offload, the context should be plain context.

hope for your help, thanks.

Best Regards,
Junhao

**Comment by paulidale** [comment]
The default was changed so that KTLS was no longer use unless it was asked for: #14799
Perhaps this is your issue?

**Comment by gaojunhao** [comment]
> The default was changed so that KTLS was no longer use unless it was asked for: #14799
> Perhaps this is your issue?

Yes, the switch of ktls enable has been changed in latest openssl, then I try the new compile command to compile nginx with openssl as follows, but ktls offload still can't work. 
nginx compile:
cd nginx; ./auto/configure --with-openssl=/path/to/openssl --with-debug --with-http_ssl_module --with-openssl-opt="enable-ktls -DSSL_OP_ENABLE_KTLS -g3"

Another question, could you share me how to debug opeenssl, i want to debug openssl by myself. thanks.

Best regards,
Junhao

**Comment by mattcaswell** [comment]
> -DSSL_OP_ENABLE_KTLS

This is not the way to enable KTLS. KTLS must be enabled both at *compile time* (using enable-ktls as you are doing), and at *run time*. The run time enabling can *only* be done by the application - so nginx will have to be changed to use the new option.

**Comment by gaojunhao** [comment]
> > -DSSL_OP_ENABLE_KTLS
> 
> This is not the way to enable KTLS. KTLS must be enabled both at _compile time_ (using enable-ktls as you are doing), and at _run time_. The run time enabling can _only_ be done by the application - so nginx will have to be changed to use the new option.

Thanks, I modified nginx source code, send ktls offload has been enabled, but recv ktls offload can't work.  I have checked issue list of nginx about ktls offload, recv ktls offload has been realized, so except enable-ktls and add SSL_OP_ENABLE_KTLS into nginx, any other operations I need to do. Thanks.

Best Regards,
Junhao

**Comment by mattcaswell** [comment]
> enable-ktls and add SSL_OP_ENABLE_KTLS into nginx, any other operations I need to do.

That should be sufficient.

**Comment by gaojunhao** [comment]
> > enable-ktls and add SSL_OP_ENABLE_KTLS into nginx, any other operations I need to do.
> 
> That should be sufficient.

Thanks for your fast reply. I have big interests in debugging openssl and want to know why recv ktls offload not work in nginx, could you share me some facts about debugging openssl :). I use BIO_printf and printf but I can't see any print on console or file.

**Comment by gaojunhao** [comment]
> > enable-ktls and add SSL_OP_ENABLE_KTLS into nginx, any other operations I need to do.
> 
> That should be sufficient.

hi mattcaswell,

My kernel version is 5.10, then I modified #define LINUX_VERSION_CODE 265728 => #define LINUX_VERSION_CODE 330240, then rx ktls related code will be executed. Then I use syslog to check code execution, find code stuck  in the following line of ssl/t1_enc.c.

count_unprocessed = count_unprocessed_records(s);

count_unprocessed is < 0, so it will goto skip_ktls. That will cause ktls recv can't be init.

Best Regards,
Junhao

**Comment by gaojunhao** [comment]
> > enable-ktls and add SSL_OP_ENABLE_KTLS into nginx, any other operations I need to do.
> 
> That should be sufficient.

I checked this notation
ifndef OPENSSL_NO_KTLS
 /*
  * Count the number of records that were not processed yet from record boundary.
  *
  * This function assumes that there are only fully formed records read in the
  * record layer. If read_ahead is enabled, then this might be false and this
  * function will fail.
  */
# ifndef OPENSSL_NO_KTLS_RX

Then I checked the value of s->rlayer.read_ahead = 1, so count_unprocessed_records(s) function will fail, then I disable the related code, run it again, but the code also stuck in the following postion of function "static ossl_inline int ktls_read_record(int fd, void *data, size_t length)".

ret = recvmsg(fd, &msg, 0);

the ret is -1, I don't know the reason of error, hope for your help, Thanks.

Best Regards,
Junhao

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
